packages/kubernetes-1.23/kubelet-exec-start-conf

```diff
@@ -38,4 +38,5 @@ ExecStart=/usr/bin/kubelet \
 {{#if settings.kubernetes.log-level includeZero=true}}
     -v {{settings.kubernetes.log-level}} \
 {{/if}}
-    --pod-infra-container-image ${POD_INFRA_CONTAINER_IMAGE}
+    --pod-infra-container-image ${POD_INFRA_CONTAINER_IMAGE} \
+    --runtime-cgroups=/runtime.slice/containerd.service
```

Comment:
This doesn't feel quite right to me. `runtime.slice` is a systemd-managed cgroup and should not be touched by any process without setting up a delegation.

The kubelet args reference says this is for:
> Optional absolute name of cgroups to create and run the runtime in.

But kubelet should be using the systemd cgroup driver and shouldn't be creating any cgroups directly.

Possibly the option has evolved and does more than what's described?

Reply:
The wording of the kubelet args ref does sound  like there’s changes being made at a system level here:

> Optional absolute name of cgroups to create and run the runtime in.

but after reading more and more issues/links related to this, it seems to me that this is meant to convey the name of the cgroup that the runtime is running in and to collect stats from. I traced this flag bag to its inception, across its old name `runtime-container`, to the commit where it all began: https://sourcegraph.com/github.com/kubernetes/kubernetes/-/commit/38efc837b906d475daf683beb559b90b9a03f5af. Here, the flag used to have the description 

> "Absolute name of the cgroups to create (if required) and run the runtime in (Default: /docker-daemon)."

Maybe that “if required” bit would make the helptext on the flag today sound less intrusive than it really is because it really will not create the cgroup if it already exists. It looks to me like a vestige of the past, when this option may actually have been creating cgroups. Even as far back as k8s 1.23, the only context in which the value(s) set by this flag are used by cadvisor: https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/kubernetes/kubernetes%24%40v1.23.0+cgrouproots&patternType=keyword&sm=0 (`cgrouproots` is populated with, among other things, the value of `--runtime-cgroups`: [here](https://sourcegraph.com/github.com/kubernetes/kubernetes@ab69524f795c42094a6630298ff53f3c3ebab7f4/-/blob/pkg/kubelet/config/flags.go?L90:18-90:32#tab=references) it is set and [here](https://sourcegraph.com/github.com/kubernetes/kubernetes@ab69524f795c42094a6630298ff53f3c3ebab7f4/-/blob/cmd/kubelet/app/server.go?L667-672) it is appended to that list)

This will be the case given the default/recommended systemd cgroup driver is used. I suppose someone out there could be using [cgroupfs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#using-the-cgroupfs-driver) but I’m not sure why they would since kubernetes does not recommend it.

I did some additional testing just to be more confident here. I created two clusters, one for k8s 1.23-1.25 nodes and one for k8s 1.26-1.29 nodes. For each node, I 1/ ensured that the change worked in viewing the runtime stats group via `kubectl get —raw "/api/v1/nodes/${NODE_NAME}/proxy/stats/summary" | jq -r '.node.systemContainers[] | .name'` and 2/ that the cgroup hierarchy via `systemd-cgls -u runtime.slice` and `systemd-cgls -u system.slice` between nodes with the change and without the change passed the “eyeball test”, aka no abnormalities.

Reply:
Great! Thank you for the extensive research and testing here. That puts my mind at ease.